feat(formatter): add monaco formatter for mdc

adamdehaven opened the pull request:
Adds a `monaco-editor` formatting provider, along with updated usage instructions and unit tests.

adamdehaven commented:
@farnabaz I have this working in my downstream project; let me know if I can make any updates or scenarios I didn't account for.

farnabaz commented:
Thanks @adamdehaven, This is great.
I tested it in the playground, but formatted does not run automatically while typing. Am I missing something?

I tried with putting `editor.getAction('editor.action.formatDocument').run()` in the editor, but in this case formatted will run once on load.

adamdehaven commented:
> I tested it in the playground, but formatted does not run automatically while typing. Am I missing something?

Ah yep @farnabaz I forgot to register the `formatOnType` provider 😅 

I have updated the formatter, playground example, and the docs to configure `formatOnType`.

When typing in the playground, it will auto-format whenever you enter a newline `\n` character. Since this impacts the editor's default indentation (e.g. keeping you at the same level as you type) I added a parameter `isFormatOnType: boolean` to the `formatter` function so that the result respects this indentation rule when typing and formats as expected when formatting the entire document.

adamdehaven commented:
I'm doing a little more refactoring to improve the formatting in some use cases, e.g. multiline strings, etc. so I changed the PR back to draft until I push a few optimizations.

farnabaz commented:
Perfect 👍 

One thing about the README.
Let's separate the use of highlighter and formatter in README to imply that they can be used separately.

adamdehaven commented:
> One thing about the README. Let's separate the use of highlighter and formatter in README to imply that they can be used separately.

Done.

Also added this note:

> [!Note]
> Since the format provider utilizes spaces for indention, we will also add a few **required** configuration properties to the Monaco Editor instance below.

farnabaz reviewed (approved):
Thanks

adamdehaven commented:
@farnabaz I saw the release went out, but there was no publish to npm?

farnabaz commented:
Strange, Thanks for update
Just released on npm
